### DISCUSSION TITLE
Mock-up HSI attribute and event messages for software development

### DISCUSSION THREAD

**Discussion by smallorange** [fwupd/fwupd#4832] [Ideas]
fwupd only can make the HSI attributes and events message based on the host system status. If the developer wants to test their program in different situations, the mock-up HSI status is necessary to fulfill every kind of scenario. The mock-up HSI message can be used to test the device security setting panel, secure boot warning for gnome-shell, Plymouth, and any program developed based on those messages. The JSON files can be used to describe the message patterns the developer wants.

The following JSON shows how we can define a mock-up message for HSI.
{
"hsi_level":1,
"SecurityAttrs":[
{
"AppstreamId":"org.fwupd.hsi.Kernel.Tainted",
"Created":1657796196,
"enabled": 0
//0 is fales and 1 is true
//optional
  "Flags": 1,
  "Guid': ["b9a2dd81-159e-5537-a7db-e7101d164d3f",
           "30249f37-d140-5d3e-9319-186b1bd5cac3",
           "a935c535-b4a9-580f-bcf3-776b64dfd328",
           "18800804-cac8-57ad-b534-fc5e4485ff94"]
}
],
"SecurityEvents":[
{
"AppstreamId":"org.fwupd.hsi.Uefi.SecureBoot",
"Created":1657269976,
"status": 0 
//0 is false and 1 is true
//optional
"Flags": 1024
}
]
}

The daemon can only output the HSI attribute and event messages based on a JSON file through org.freedesktop.fwupd.GetHostSecurityAttrs() and org.freedesktop.fwupd.GetHostSecurityEvents(). To keep the JSON simple, only AppStreamID, time, and status are required. The remaining items, such as Flags, GUID are optional items. The HSI mockup feature could be a new command line daemon or embedded in fwupd and we can choose one of the system architectures to implement this feature. I believe this is good for every project development which needs firmware security information.

Thank you :)

**Comment by hughsie** [comment]
I think this is an excellent idea; I also think it makes a lot of sense to mock up lists of devices too (e.g. to test gnome-firmware), although maybe allowing those device to be updatable might be one step too far.

I think using the output of `fwupdmgr security --json` makes sense, although I'm not sure where we should store the test data -- perhaps `/usr/share/fwupd/test-profiles.d/foo.json` or something, and then have the ability to load that profile at startup by doing something like `FWUPD_TEST_PROFILE=foo ./src/fwupd --verbose` etc.

We already have `fu_security_attrs_from_json()` (thanks to you!) so I think it should be pretty simple to plumb into fwupd. @superm1 any reservations?
